osu.Game/Screens/Play/HUD/PerformancePointsCounter.cs

```diff
@@ -0,0 +1,173 @@
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.Annotations;
+using osu.Framework.Allocation;
+using osu.Framework.Audio.Track;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Graphics.Textures;
+using osu.Framework.Localisation;
+using osu.Game.Beatmaps;
+using osu.Game.Graphics;
+using osu.Game.Graphics.Sprites;
+using osu.Game.Graphics.UserInterface;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Difficulty;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Scoring;
+using osu.Game.Skinning;
+using osuTK;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    public class PerformancePointsCounter : RollingCounter<int>, ISkinnableDrawable
+    {
+        public bool UsesFixedAnchor { get; set; }
+
+        [CanBeNull]
+        [Resolved(CanBeNull = true)]
+        private ScoreProcessor scoreProcessor { get; set; }
+
+        [Resolved(CanBeNull = true)]
+        [CanBeNull]
+        private GameplayState gameplayState { get; set; }
+
+        [CanBeNull]
+        private TimedDifficultyAttributes[] timedAttributes;
+
+        [CanBeNull]
+        private Ruleset gameplayRuleset;
+
+        private readonly CancellationTokenSource loadCancellationSource = new CancellationTokenSource();
+
+        public PerformancePointsCounter()
+        {
+            Current.Value = DisplayedCount = 0;
+        }
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours, BeatmapDifficultyCache difficultyCache)
+        {
+            Colour = colours.BlueLighter;
+
+            if (gameplayState != null)
+            {
+                gameplayRuleset = gameplayState.Ruleset;
+                difficultyCache.GetTimedDifficultyAttributesAsync(new GameplayWorkingBeatmap(gameplayState.Beatmap), gameplayRuleset, gameplayState.Mods.ToArray(), loadCancellationSource.Token)
+                               .ContinueWith(r => Schedule(() => timedAttributes = r.Result), TaskContinuationOptions.OnlyOnRanToCompletion);
+            }
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            if (scoreProcessor != null)
+                scoreProcessor.NewJudgement += onNewJudgement;
+        }
+
+        private void onNewJudgement(JudgementResult judgement)
+        {
+            if (gameplayState?.Score == null || timedAttributes == null || timedAttributes.Length == 0)
+                return;
+
+            Debug.Assert(gameplayRuleset != null);
+
+            var attribIndex = Array.BinarySearch(timedAttributes, 0, timedAttributes.Length, new TimedDifficultyAttributes(judgement.HitObject.GetEndTime(), null));
+            if (attribIndex < 0)
+                attribIndex = ~attribIndex - 1;
+            attribIndex = Math.Clamp(attribIndex, 0, timedAttributes.Length - 1);
```

Comment:
is it actually possible for a value to be outside this range spec?

Reply:
Yes. Specifically the max value.

> If this index is one greater than the upper bound of the array, there are no elements larger than value in the array

Meaning, `~attribIndex - 1` brings the value to `timedAttributes.Length`.